StripeCardScan/StripeCardScan/Source/CardScan/UI/ScanBaseViewController.swift

```diff
@@ -413,22 +413,20 @@ class ScanBaseViewController: UIViewController, AVCaptureVideoDataOutputSampleBu
         from connection: AVCaptureConnection
     ) {
         if self.machineLearningSemaphore.wait(timeout: .now()) == .success {
-            ScanBaseViewController.machineLearningQueue.async {
-                self.captureOutputWork(sampleBuffer: sampleBuffer)
+            guard let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer),
+                  let fullCameraImage = pixelBuffer.cgImage() else {
                 self.machineLearningSemaphore.signal()
+                return
             }
-        }
-    }
-
-    func captureOutputWork(sampleBuffer: CMSampleBuffer) {
-        guard let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else {
-            return
-        }
 
-        guard let fullCameraImage = pixelBuffer.cgImage() else {
-            return
+            ScanBaseViewController.machineLearningQueue.async { [weak self] in
+                self?.captureOutputWork(fullCameraImage: fullCameraImage)
+                self?.machineLearningSemaphore.signal()
+            }
         }
+    }
```

Comment:
Based on my memory from implementing the equivalent for Identity, `captureOutput` on the `AVCaptureVideoDataOutputSampleBufferDelegate` is always called off the main thread on its own dedicated thread, so I think this should be okay.

Reply:
* I think it might be a little expensive, but I agree that I don't think this will be called on the main thread
* I don't have the full logic downloaded into my brain atm, but is it okay to call the `machineLearningSemaphore` outside the machineLearningQueue? That would free anyone else `wait`-ing on it, right?

Reply:
> is it okay to call the machineLearningSemaphore outside the machineLearningQueue?

Yes. In fact, this is what it's designed for. Generally the camera feed is faster than the ML model can process images, so the idea is that you want to block the thread that is reading from the camera in order to finish processing the image in CoreML before asking for more camera frames.

Here's an excerpt from the copy of the CoreML survival guide that gives a basic example:
https://drive.google.com/open?id=1gwW8pgJvg8t_EOkM2cgnDsQDI5Hc_f4o&disco=AAAAVxpjn24

Public link (page 400):
https://www.scribd.com/document/689236733/Core-ML-Survival-Guide

Reply:
So yeah, given that we're explicitly calling `wait` in order to pause this DispatchQueue, I think that it's inconsequential that converting `CMSampleBuffer` -> `CIImage` could be non-performant